### PR TITLE
SDSTOR-22421: shard superblk backward compat for log-only shard refactor

### PR DESCRIPTION
**Pull request by Hooper9973**
- Migrate v2 shard superblk (sb_version=0x02) to v3 on recovery; unify decode logic into decode_shard_sb() shared by raft-log path and on_shard_meta_blk_found

- Decode old shard raft-log messages (v2 header_extn) on new code path; extend ResyncShardMetaData schema with sealed_lsn at end for old-buffer compat

- Override get_blk_alloc_hint for CREATE/SEAL_SHARD to return committed-blk hint, skipping data-block allocation for log-only shard messages

### CODE REVIEW

File: src/lib/homestore_backend/hs_shard_manager.cpp [hunk 1]
```diff
@@ -26,6 +26,37 @@ SISL_LOGGING_DECL(shardmgr)
...
+    info.current_leader = v2.current_leader;
+    std::memcpy(info.meta, v2.meta, ShardInfo::meta_length);
+    return info;
+}
+
+// Core decode: given a typed shard_info_superblk pointer, extract ShardInfo + chunk IDs.
+// Handles all supported on-disk versions. Returns nullopt for unknown versions.
+static std::optional< std::tuple< ShardInfo, homestore::chunk_num_t, homestore::chunk_num_t > >
+decode_shard_sb(const HSHomeObject::shard_info_superblk* sb) {
+    if (sb->sb_version >= HSHomeObject::shard_info_superblk::shard_sb_version) {
```
Comment (xiaoxichen) [inline src/lib/homestore_backend/hs_shard_manager.cpp]: what would be the situation of `sb->sb_version > HSHomeObject::shard_info_superblk::shard_sb_version` ? older version read data written by newer version?

File: src/include/homeobject/shard_manager.hpp [hunk 2]
```diff
@@ -35,14 +35,14 @@ struct ShardInfo {
     shard_id_t id;
     pg_id_t placement_group;
     State state;
-    uint64_t lsn;                   // created_lsn
-    uint64_t sealed_lsn{INT64_MAX}; // sealed_lsn
+    uint64_t lsn; // created_lsn
```
Comment (xiaoxichen) [inline src/include/homeobject/shard_manager.hpp]: maybe rename it as create_lsn?

**Review by xiaoxichen — COMMENTED** [review]
lgtm in general

File: src/lib/homestore_backend/replication_state_machine.cpp [hunk 3]
```diff
@@ -213,13 +212,21 @@ ReplicationStateMachine::get_blk_alloc_hints(sisl::blob const& header, uint32_t
...
+        if (data_size == 0) {
+            // New log-only format: should not reach here, but handle gracefully.
+            LOGW("get_blk_alloc_hints called for log-only shard message type={}, shard={}, pg={}", msg_header->msg_type,
+                 msg_header->shard_id, msg_header->pg_id);
+            return homestore::blk_alloc_hints{};
+        } else {
+            // Old format: shard message carried shard header/footer data blocks. Return committed_blk_id
+            // so HomeStore skips block allocation and data write. Shard on_commit will ignore pbas.
+            homestore::blk_alloc_hints hints;
+            hints.committed_blk_id = homestore::MultiBlkId{homestore::BlkId{0, 1, 0xFFFF}};
```
Comment (xiaoxichen) [inline src/lib/homestore_backend/replication_state_machine.cpp]: make this as a global constant (like tombstone_pba) so other pieces of code is easier to check/assert based on this value.

File: src/lib/homestore_backend/tests/hs_shard_tests.cpp [hunk 4]
```diff
@@ -370,35 +372,37 @@ TEST_F(HomeObjectFixture, ShardVersionMigrationRecovery) {
             old_sb->info.available_capacity_bytes = orig_data.info.available_capacity_bytes;
             old_sb->info.total_capacity_bytes = orig_data.info.total_capacity_bytes;
             old_sb->info.current_leader = orig_data.info.current_leader;
-            // Note: v1 doesn't have meta field, so we don't copy it
+            std::memcpy(old_sb->info.meta, orig_data.info.meta, ShardInfo::meta_length);
+            // Note: sealed_lsn is not present in v2, so we don't set it
             old_sb->p_chunk_id = orig_data.p_chunk_id;
             old_sb->v_chunk_id = orig_data.v_chunk_id;
             old_sb.write();
+            LOGINFO("Created v2 shard {}", shard_id);
```
Comment (JacksonYao287) [inline src/lib/homestore_backend/tests/hs_shard_tests.cpp]: remove this log, since line 363 has already loged

File: src/lib/homestore_backend/replication_state_machine.cpp [hunk 5]
```diff
@@ -213,13 +212,21 @@ ReplicationStateMachine::get_blk_alloc_hints(sisl::blob const& header, uint32_t
     switch (msg_header->msg_type) {
     case ReplicationMessageType::CREATE_SHARD_MSG:
     case ReplicationMessageType::SEAL_SHARD_MSG: {
```
Comment (JacksonYao287) [inline src/lib/homestore_backend/replication_state_machine.cpp:214]: suggest adding a release_assert(msg_header->msg_type != ReplicationMessageType::CREATE_SHARD_MSG && msg_header->msg_type != ReplicationMessageType::SEAL_SHARD_MSG)

IIUC， it is not expected that new version HO receives a old-version log?

Comment (xiaoxichen) [reply src/lib/homestore_backend/replication_state_machine.cpp:214]: it can happen... for example one member is upgraded and leader(not upgraded) sends log to it.

Comment (JacksonYao287) [reply src/lib/homestore_backend/replication_state_machine.cpp:214]: in my mind, we will upgrade sm to this new version subcluster by subcluster, and this is a non-rolling upgrade. 

In my mind, for each subcluster :
1 stop writing to this subcluster.
2 check if all the sms in this cluster have committed to the last log in log store.
3 check if all the last commit log of sms in the same pg are identical. 

the aim of above steps is to guarantee that no new logs will be committed and no new logs will be sent from leader to follower.

now, we can safely do the upgrade in this subcluster.

Comment (xiaoxichen) [reply src/lib/homestore_backend/replication_state_machine.cpp:214]: as we cannot do the CR,  the step above is very tricky to be executed.  
we can disable this SC from DS so that no more create_shard will happen on this SC.
we can even disable DM, it is a config map update, which stops the create shard/seal shard completely.
but we cannot stop write easily as gateway can still writes to open shard... we cannot stop delete-blob.

with that , no way to guarantee no writes to this SC. your step 2/3 is tricky to execute.

Comment (xiaoxichen) [reply src/lib/homestore_backend/replication_state_machine.cpp:214]: it is not 2/3 is impossible,  just to minimize the upgrade effort, if some cases can be handled in code without too much corner cases.

File: src/lib/homestore_backend/tests/hs_shard_tests.cpp (hunk 4 above)
Comment (Hooper9973) [reply src/lib/homestore_backend/tests/hs_shard_tests.cpp]: since 363 is removed, this line will be retained.

File: src/lib/homestore_backend/hs_shard_manager.cpp (hunk 1 above)
Comment (Hooper9973) [reply src/lib/homestore_backend/hs_shard_manager.cpp]: `sb_version > shard_sb_version` happens when older SM receive new SM's msg (rolling upgrade). With >=, old code silently parses unknown future struct layout — dangerous if new version changed field offsets or types. Returning nullopt here causes RELEASE_ASSERT at callers, which is more safe than slient unexpected error. 
Changed to ==.

File: src/lib/homestore_backend/hs_shard_manager.cpp [hunk 6]
```diff
@@ -584,56 +623,39 @@ void HSHomeObject::on_shard_message_commit(int64_t lsn, sisl::blob const& h, sha
 }
 
 void HSHomeObject::on_shard_meta_blk_found(homestore::meta_blk* mblk, sisl::byte_view buf) {
-    // First peek at the version
     auto* header = reinterpret_cast< const DataHeader* >(buf.bytes());
+    RELEASE_ASSERT(header->version == DataHeader::data_header_version, "Unknown shard superblock DataHeader version {}",
```
Comment (Besroy) [inline src/lib/homestore_backend/hs_shard_manager.cpp]: Could you use `valid()` here, or add a clear marker so this if can be removed when we upgrade the data version later?  This assert will likely fail during future data version updates, and the original context may be lost by then.

File: src/lib/homestore_backend/replication_state_machine.cpp (hunk 5 above)
Comment (JacksonYao287) [reply src/lib/homestore_backend/replication_state_machine.cpp:214]: ok.   I want to do this because assert can helping us identify the unexpected case after upgrade is done. 

we can change this part in the next release after upgrade is done.

File: src/lib/homestore_backend/replication_state_machine.cpp (hunk 3 above)
Comment (Hooper9973) [reply src/lib/homestore_backend/replication_state_machine.cpp]: changed it to tombstone_pba directly.
Although chunk_id seems hardly to be the 0xFFFF, but tombstone_pba is more safe.
And when rollback mesg, homestore will free the blk, and tombstone_pba's blk_count is 0 which is more safe.

File: src/lib/homestore_backend/hs_shard_manager.cpp (hunk 6 above)
Comment (Hooper9973) [reply src/lib/homestore_backend/hs_shard_manager.cpp]: Good!

File: src/include/homeobject/shard_manager.hpp (hunk 2 above)
Comment (Hooper9973) [reply src/include/homeobject/shard_manager.hpp]: Done

File: src/lib/homestore_backend/tests/hs_shard_tests.cpp [hunk 7]
```diff
@@ -321,84 +321,87 @@ TEST_F(HomeObjectFixture, ShardVersionMigrationRecovery) {
...
-            old_sb->version = 0x01; // Old version
-            // Simulate the actual bug: first v1 shard has incorrect type field set to BLOB_INFO
+            old_sb->version = HSHomeObject::DataHeader::data_header_version;
+            old_sb->type = HSHomeObject::DataHeader::data_type_t::SHARD_INFO;
+            old_sb->sb_version = 0x02;
+            // Simulate the actual bug: first v2 shard has incorrect type field set to BLOB_INFO
             if (shard_id == shard_ids[0]) {
                 old_sb->type = HSHomeObject::DataHeader::data_type_t::BLOB_INFO; // Bug scenario
-                LOGINFO("Created v1 shard {} with BLOB_INFO type (bug scenario)", shard_id);
+                LOGINFO("Created v2 shard {} with BLOB_INFO type (bug scenario)", shard_id);
```
Comment (koujl) [inline src/lib/homestore_backend/tests/hs_shard_tests.cpp]: The bug is already fixed, let's remove this part.

Comment (Hooper9973) [reply src/lib/homestore_backend/tests/hs_shard_tests.cpp]: done

**Review by JacksonYao287 — APPROVED** [review]
LGTM
